// nestjs-trpc@1.4.0/dist/trpc.module.js
"use strict";
var TRPCModule_1;
Object.defineProperty(exports, "__esModule", { value: true });
exports.TRPCModule = void 0;
const tslib_1 = require("tslib");
const common_1 = require("@nestjs/common");
const core_1 = require("@nestjs/core");
const trpc_constants_1 = require("./trpc.constants");
const trpc_driver_1 = require("./trpc.driver");
const trpc_factory_1 = require("./factories/trpc.factory");
const router_factory_1 = require("./factories/router.factory");
const procedure_factory_1 = require("./factories/procedure.factory");
const middleware_factory_1 = require("./factories/middleware.factory");
const trpc_generator_1 = require("./generators/trpc.generator");
const decorator_generator_1 = require("./generators/decorator.generator");
const router_generator_1 = require("./generators/router.generator");
const middleware_generator_1 = require("./generators/middleware.generator");
const context_generator_1 = require("./generators/context.generator");
const app_router_host_1 = require("./app-router.host");
const drivers_1 = require("./drivers");
let TRPCModule = TRPCModule_1 = class TRPCModule {
    static forRoot(options = {}) {
        return {
            module: TRPCModule_1,
            providers: [{ provide: trpc_constants_1.TRPC_MODULE_OPTIONS, useValue: options }],
        };
    }
    async onModuleInit() {
        const httpAdapter = this.httpAdapterHost?.httpAdapter;
        if (!httpAdapter) {
            return;
        }
        this.consoleLogger.setContext(trpc_constants_1.LOGGER_CONTEXT);
        await this.trpcDriver.start(this.options);
        const platformName = httpAdapter.getType();
        if (this.appRouterHost.appRouter != null) {
            this.consoleLogger.log(`Server has been initialized successfully using the ${platformName} driver.`, 'TRPC Server');
        }
    }
};
exports.TRPCModule = TRPCModule;
tslib_1.__decorate([
    (0, common_1.Inject)(trpc_constants_1.TRPC_MODULE_OPTIONS),
    tslib_1.__metadata("design:type", Object)
], TRPCModule.prototype, "options", void 0);
tslib_1.__decorate([
    (0, common_1.Inject)(common_1.ConsoleLogger),
    tslib_1.__metadata("design:type", common_1.ConsoleLogger)
], TRPCModule.prototype, "consoleLogger", void 0);
tslib_1.__decorate([
    (0, common_1.Inject)(core_1.HttpAdapterHost),
    tslib_1.__metadata("design:type", core_1.HttpAdapterHost)
], TRPCModule.prototype, "httpAdapterHost", void 0);
tslib_1.__decorate([
    (0, common_1.Inject)(trpc_driver_1.TRPCDriver),
    tslib_1.__metadata("design:type", trpc_driver_1.TRPCDriver)
], TRPCModule.prototype, "trpcDriver", void 0);
tslib_1.__decorate([
    (0, common_1.Inject)(app_router_host_1.AppRouterHost),
    tslib_1.__metadata("design:type", app_router_host_1.AppRouterHost)
], TRPCModule.prototype, "appRouterHost", void 0);
exports.TRPCModule = TRPCModule = TRPCModule_1 = tslib_1.__decorate([
    (0, common_1.Module)({
        imports: [],
        providers: [
            common_1.ConsoleLogger,
            trpc_driver_1.TRPCDriver,
            trpc_factory_1.TRPCFactory,
            core_1.MetadataScanner,
            router_factory_1.RouterFactory,
            procedure_factory_1.ProcedureFactory,
            middleware_factory_1.MiddlewareFactory,
            decorator_generator_1.DecoratorGenerator,
            middleware_generator_1.MiddlewareGenerator,
            context_generator_1.ContextGenerator,
            router_generator_1.RouterGenerator,
            trpc_generator_1.TRPCGenerator,
            app_router_host_1.AppRouterHost,
            drivers_1.FastifyDriver,
            drivers_1.ExpressDriver,
        ],
        exports: [app_router_host_1.AppRouterHost],
    })
], TRPCModule);
