// nestjs-trpc@1.4.0/dist/generators/context.generator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ContextGenerator = void 0;
const tslib_1 = require("tslib");
const ts_morph_1 = require("ts-morph");
const common_1 = require("@nestjs/common");
const func_loc_1 = require("func-loc");
let ContextGenerator = class ContextGenerator {
    async getContextInterface(context, project) {
        const className = context?.name;
        if (!className) {
            return null;
        }
        const contextInstance = new context();
        if (typeof contextInstance.create !== 'function') {
            return null;
        }
        const contextFileLocation = await (0, func_loc_1.locate)(contextInstance.create, {
            sourceMap: true,
        });
        const contextSourceFile = project.addSourceFileAtPath(contextFileLocation.path);
        const classDeclaration = this.getClassDeclaration(contextSourceFile, context.name);
        if (!classDeclaration) {
            return null;
        }
        const createMethod = classDeclaration.getMethod('create');
        if (!createMethod) {
            return null;
        }
        const ctxType = this.extractReturnTypeFromCreateMethod(createMethod);
        if (!ctxType) {
            return null;
        }
        return ctxType.getText();
    }
    extractReturnTypeFromCreateMethod(createMethod) {
        const body = createMethod.getBody();
        if (!body)
            return null;
        // Find the return statement
        const returnStatement = body
            .getDescendantsOfKind(ts_morph_1.SyntaxKind.ReturnStatement)
            .find((statement) => statement.getExpression() !== undefined);
        if (!returnStatement)
            return null;
        const returnExpression = returnStatement.getExpression();
        if (!returnExpression)
            return null;
        // Get the type of the returned expression
        const returnType = returnExpression.getType();
        // Check if the type is a Promise
        if (this.isPromiseType(returnType)) {
            // Get the type argument of the Promise
            const typeArguments = returnType.getTypeArguments();
            return typeArguments.length > 0 ? typeArguments[0] : null;
        }
        return returnType;
    }
    isPromiseType(type) {
        return (type.getSymbol()?.getName() === 'Promise' ||
            type.getSymbol()?.getName() === '__global.Promise' ||
            type.getText().startsWith('Promise<'));
    }
    getClassDeclaration(sourceFile, className) {
        const classDeclaration = sourceFile.getClass(className);
        if (classDeclaration) {
            return classDeclaration;
        }
        return undefined;
    }
};
exports.ContextGenerator = ContextGenerator;
exports.ContextGenerator = ContextGenerator = tslib_1.__decorate([
    (0, common_1.Injectable)()
], ContextGenerator);
