// nestjs-trpc@1.6.1/dist/factories/router.factory.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RouterFactory = void 0;
const tslib_1 = require("tslib");
const common_1 = require("@nestjs/common");
const core_1 = require("@nestjs/core");
const lodash_1 = require("lodash");
const trpc_constants_1 = require("../trpc.constants");
const procedure_factory_1 = require("./procedure.factory");
let RouterFactory = class RouterFactory {
    getRouters() {
        const routers = [];
        this.modulesContainer.forEach((moduleRef) => {
            moduleRef.providers.forEach((wrapper) => {
                const router = this.extractRouterFromWrapper(wrapper);
                if (router != null) {
                    routers.push(router);
                }
            });
        });
        return routers;
    }
    extractRouterFromWrapper(wrapper) {
        const { instance, name } = wrapper;
        if (instance == null) {
            return null;
        }
        const router = Reflect.getMetadata(trpc_constants_1.ROUTER_METADATA_KEY, instance.constructor);
        if (router == null) {
            return null;
        }
        const middlewares = Reflect.getMetadata(trpc_constants_1.MIDDLEWARES_KEY, instance.constructor) || [];
        return {
            name,
            instance,
            path: router.path,
            alias: router.alias,
            middlewares: middlewares,
        };
    }
    serializeRoutes(router, procedure) {
        const routers = this.getRouters();
        const routerSchema = Object.create({});
        routers.forEach((route) => {
            const { instance, name, middlewares, alias } = route;
            const camelCasedRouterName = (0, lodash_1.camelCase)(alias ?? name);
            const prototype = Object.getPrototypeOf(instance);
            const procedures = this.procedureFactory.getProcedures(instance, prototype);
            this.consoleLogger.log(`Router ${name} as ${camelCasedRouterName}.`, 'Router Factory');
            const routerProcedures = this.procedureFactory.serializeProcedures(procedures, instance, camelCasedRouterName, procedure, middlewares);
            // TODO: To get this working with `trpc` v11, we need to remove the `router()` method from here.
            routerSchema[camelCasedRouterName] = router(routerProcedures);
        });
        return routerSchema;
    }
};
exports.RouterFactory = RouterFactory;
tslib_1.__decorate([
    (0, common_1.Inject)(common_1.ConsoleLogger),
    tslib_1.__metadata("design:type", common_1.ConsoleLogger)
], RouterFactory.prototype, "consoleLogger", void 0);
tslib_1.__decorate([
    (0, common_1.Inject)(core_1.ModulesContainer),
    tslib_1.__metadata("design:type", core_1.ModulesContainer)
], RouterFactory.prototype, "modulesContainer", void 0);
tslib_1.__decorate([
    (0, common_1.Inject)(procedure_factory_1.ProcedureFactory),
    tslib_1.__metadata("design:type", procedure_factory_1.ProcedureFactory)
], RouterFactory.prototype, "procedureFactory", void 0);
exports.RouterFactory = RouterFactory = tslib_1.__decorate([
    (0, common_1.Injectable)()
], RouterFactory);
